ReactiveCocoa/Swift/SignalProducer.swift

```diff
@@ -1674,172 +1672,188 @@ extension SignalProducerProtocol {
-			let shouldStartUnderlyingProducer: Bool
-
-			lock.lock()
-			if let producer = producer, let producerObserver = producerObserver {
-				(initializedProducer, initializedObserver) = (producer, producerObserver)
-				shouldStartUnderlyingProducer = false
-			} else {
-				let (producerTemp, observerTemp) = SignalProducer<Value, Error>.bufferingProducer(upTo: capacity)
+		let state = Atomic(ReplayState<Value, Error>(upTo: capacity))
 
-				(producer, producerObserver) = (producerTemp, observerTemp)
-				(initializedProducer, initializedObserver) = (producerTemp, observerTemp)
-				shouldStartUnderlyingProducer = true
-			}
-			lock.unlock()
-
-			// subscribe `observer` before starting the underlying producer.
-			disposable += initializedProducer.start(observer)
-			disposable += {
-				// Don't dispose of the original producer until all observers
-				// have terminated.
-				lifetime = nil
-			}
-
-			if shouldStartUnderlyingProducer {
-				self.take(during: lifetime!)
-					.start(initializedObserver)
-			}
-		}
-	}
-}
-
-extension SignalProducer {
-	private static func bufferingProducer(upTo capacity: Int) -> (SignalProducer, Signal<Value, Error>.Observer) {
-		precondition(capacity >= 0, "Invalid capacity: \(capacity)")
-
-		// Used as an atomic variable so we can remove observers without needing
-		// to run on a serial queue.
-		let state: Atomic<BufferState<Value, Error>> = Atomic(BufferState())
-
-		let producer = self.init { observer, disposable in
-			// Assigned to when replay() is invoked synchronously below.
-			var token: RemovalToken?
-
-			let replayBuffer = ReplayBuffer<Value>()
-			var replayValues: [Value] = []
-			var replayToken: RemovalToken?
-			var next = state.modify { state in
-				replayValues = state.values
-				if replayValues.isEmpty {
-					token = state.observers?.insert(observer)
-				} else {
-					replayToken = state.replayBuffers.insert(replayBuffer)
+		let bootstrap = ActionDisposable { [ended = lifetime.ended] in
+			// Start the underlying producer.
+			self
+				.take(until: ended)
+				.start { event in
+					let originalState = state.modify { state in
+						state.enqueue(event)
+					}
+					originalState.broadcast(event)
 				}
-			}
+		}
 
-			while !replayValues.isEmpty {
-				replayValues.forEach(observer.sendNext)
+		return SignalProducer { observer, disposable in
+			// Don't dispose of the original producer until all observers
+			// have terminated.
+			disposable += { _ = lifetime }
+
+			repeat {
+				do {
+					var token: RemovalToken?
+					try state.modify { state in
+						token = try state.observe(observer)
+					}
 
-				next = state.modify { state in
-					replayValues = replayBuffer.values
-					replayBuffer.values = []
-					if replayValues.isEmpty {
-						if let replayToken = replayToken {
-							state.replayBuffers.remove(using: replayToken)
+					if let token = token {
+						disposable += {
+							state.modify { state in
+								state.removeObserver(using: token)
```

Comment:
We preferred named parameters for `Atomic.modify` in the rest of the codebase though. Should we change those at some point too?